examples/reaction-rate-learning/reaction-rate-learning.py

```diff
@@ -0,0 +1,317 @@
+# and computes the loss for each time step, if a loss function (`loss_func`) is provided:
+
+from fenics import *
+set_log_level(50)
+
+def solve_reaction_system(mesh, T, num_steps, reaction_func, loss_func=lambda n,x: 0):
+
+    dt = T / num_steps # time step size
+    eps = 0.01         # diffusion coefficient
+
+    # Define function space for velocity
+    W = VectorFunctionSpace(mesh, 'CG', 2)
+
+    # Define function space for system of concentrations
+    P1 = FiniteElement('CG', triangle, 1)
+    element = MixedElement([P1, P1, P1])
+    V = FunctionSpace(mesh, element)
+
+    # Define test functions
+    v_1, v_2, v_3 = TestFunctions(V)
+
+    # Define functions for velocity and concentrations
+    w = Function(W)
+    u = Function(V)
+    u_n = Function(V)
+
+    # Split system functions to access components
+    u_1, u_2, u_3 = split(u)
+    u_n1, u_n2, u_n3 = split(u_n)
+
+    # Define source terms
+    f_1 = Expression('pow(x[0]-0.1,2)+pow(x[1]-0.1,2)<0.05*0.05 ? 10 : 0',
+                     degree=1)
+    f_2 = Expression('pow(x[0]-0.1,2)+pow(x[1]-0.3,2)<0.05*0.05 ? 10 : 0',
+                     degree=1)
+    f_3 = Constant(0)
+
+    # Define expressions used in variational forms
+    k = Constant(dt)
+    eps = Constant(eps)
+
+    # Define variational problem
+    F = (((u_1 - u_n1) / k)*v_1*dx + dot(w, grad(u_1))*v_1*dx 
+      + eps*dot(grad(u_1), grad(v_1))*dx + reaction_func(u_1, u_2)*v_1*dx  
+      + ((u_2 - u_n2) / k)*v_2*dx + dot(w, grad(u_2))*v_2*dx 
+      + eps*dot(grad(u_2), grad(v_2))*dx + reaction_func(u_1, u_2)*v_2*dx  
+      + ((u_3 - u_n3) / k)*v_3*dx + dot(w, grad(u_3))*v_3*dx 
+      + eps*dot(grad(u_3), grad(v_3))*dx - reaction_func(u_1, u_2)*v_3*dx 
+      - f_1*v_1*dx - f_2*v_2*dx - f_3*v_3*dx)
+
+    # Create time series for reading velocity data
+    timeseries_w = TimeSeries('navier_stokes_cylinder/velocity_series')
+    timeseries_w.retrieve(w.vector(), 2.0)
+
+    # Time-stepping
+    t = 0
+    results = []
+    loss = 0.0
+    for n in range(num_steps):
+
+        # Update current time
+        t += dt
+
+        # Solve variational problem for time step
+        solve(F == 0, u)
+
+        # Save solution to file (VTK)
+        _u_1, _u_2, _u_3 = u.split()
+        _u_1.rename("u1","u1")
+        _u_2.rename("u2","u3")
+        _u_3.rename("u3","u3")
+
+        # Update previous solution
+        u_n.assign(u)
+        loss += loss_func(n, u.split())
+        results.append(u.copy())
+
+    return loss, results
+
+# For the NN part, we rely on the 
+# `NN implementation <https://github.com/sebastkm/hybrid-fem-nn/blob/master/neural_network.py>`_
+# by :cite:`mitusch2021hybrid`.
+
+# Putting those two components together, we can define the hybrid PDE-NN
+# training problem in a few lines of code. First, we generate the ground
+# truth training data with the reaction rate :math:`R(u_1, u_2)= K u_1 u_2`:
+
+from neural_network import *
```

Comment:
I would change this to `from neural_network import ANN`. Only `ANN` is needed and it becomes clear what this actually imports.